probe-rs/src/architecture/arm/communication_interface.rs

```diff
@@ -329,10 +337,33 @@ pub struct MemoryApInformation {
 /// (E.g. CMSIS-DAP and J-Link support this, ST-Link does not)
 #[derive(Debug)]
 pub struct ArmCommunicationInterface<S: ArmDebugState> {
-    probe: Box<dyn DapProbe>,
+    probe: Option<Box<dyn DapProbe>>,
     state: S,
 }
 
+impl<S: ArmDebugState> Drop for ArmCommunicationInterface<S> {
+    fn drop(&mut self) {
+        if let Some(mut probe) = self.probe.take() {
+            self.state.disconnect(&mut *probe);
+        }
+    }
+}
+
+impl<S: ArmDebugState> ArmCommunicationInterface<S> {
+    fn probe_mut(&mut self) -> &mut dyn DapProbe {
+        // Unwrap: Probe is only taken when the struct is dropped
+        &mut *self.probe.as_mut().expect("ArmCommunicationInterface is in an inconsistent state. This is a bug, please report it.").as_mut()
```

Comment:
```suggestion
        self.probe.as_deref().expect("ArmCommunicationInterface is in an inconsistent state. This is a bug, please report it.")
```

Wouldn't this work, too? Did a bit of experimenting on [godbolt.org](https://rust.godbolt.org/z/zv4Ma5cTq) but right now I'm not sure I managed to grab the problem

Reply:
`as_deref_mut()`, thanks for the hint.

Reply:
> Wouldn't this work, too? Did a bit of experimenting on [godbolt.org](https://rust.godbolt.org/z/zv4Ma5cTq) but right now I'm not sure I managed to grasp the problem correctly

You can also try the code space on Github for things like this, there's a button next to the PR title.